File: backend/app/models/contest.py
from beanie import Document
from pydantic import Field
from datetime import datetime
from typing import List, Optional

class Contest(Document):
    id: str = Field(default_factory=lambda: "", alias="_id")
    title: str
    start_time: datetime
    end_time: datetime
    duration_minutes: int = 120
    participant_count: int = 0
    max_score: int = 0
    is_leaderboard_frozen: bool = False
    freeze_time_remaining_minutes: Optional[int] = None
    created_by: Optional[str] = None
    problem_ids: List[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "contests"

File: backend/src/db/models/Contest.ts
import { Schema, model } from 'mongoose';

export interface IContest {
  _id: string;
  title: string;
  startTime: Date;
  endTime: Date;
  durationMinutes: number;
  participantCount: number;
  maxScore: number;
  isLeaderboardFrozen: boolean;
  freezeTimeRemainingMinutes?: number | null;
  createdBy?: string;
  problemIds: string[];
  createdAt: Date;
}

const contestSchema = new Schema<IContest>(
  {
    _id: { type: String, required: true },
    title: { type: String, required: true },
    startTime: { type: Date, required: true },
    endTime: { type: Date, required: true },
    durationMinutes: { type: Number, default: 120 },
    participantCount: { type: Number, default: 0 },
    maxScore: { type: Number, default: 0 },
    isLeaderboardFrozen: { type: Boolean, default: false },
    freezeTimeRemainingMinutes: { type: Number, default: null },
    createdBy: { type: String, ref: 'User' },
    problemIds: [{ type: String, ref: 'Problem' }],
    createdAt: { type: Date, default: Date.now },
  },
  {
    timestamps: false,
    versionKey: false,
  },
);

export const Contest = model<IContest>('Contest', contestSchema);
